R/utils-cut.R

```diff
@@ -44,12 +44,25 @@ cut_width <- function(x, width, center = NULL, boundary = NULL, closed = c("righ
 
   # Determine bins
   min_x <- find_origin(x_range, width, boundary)
-  # Small correction factor so that we don't get an extra bin when, for
-  # example, origin = 0, max(x) = 20, width = 10.
-  max_x <- max(x, na.rm = TRUE) + (1 - 1e-08) * width
+  max_x <- max(x, na.rm = TRUE)
 
-  breaks <- c(seq(min_x, max_x, width), max_x)
-  cut(x, breaks, include.lowest = TRUE, right = (closed == "right"), ...)
+  breaks <- c(seq(min_x, max_x, width))
+  # Round breaks *before* they go into the cut function. This way the data (not rounded)
+  # will be correctly divided into the rounded bins
+  if (all(x %% 1 == 0)) {
+    avgDigits <- 0
+  } else {
+    avgDigits <- floor(mean(stringi::stri_count_regex(as.character(x), "[[:digit:]]")))
+  }
+  breaks <- as.numeric(formatC(0 + breaks, digits = avgDigits, width = 1L))
+
+  # If now, after rounded, our max bin does not include the max of the data, add another bin
```

Comment:
Is this just for safety?

Is modifying the final break a viable alternative that keeps the same number of desired bins (but the last one not officially the same width, so problematic, I guess)?

```
  if (max(breaks) < max_x) {
    breaks[length(breaks)] <- max_x
  }
```

Reply:
oooh that's an interesting idea. Because we're rounding breakpoints before cutting we're already making the bins slightly different widths 😅 The fear would be that `cut` does it's own rounding. We round the bin labels before sending to `cut` to try to force it to not round. If it suddenly sees one breakpoint with a bunch of decimals and everything else with only 2 decimals, then it might round that last breakpoint again anyways.

But what we could do instead is something like 
```
if (max(breaks) < max_x) {
  breaks[length(breaks)] <- breaks[length(breaks)]  + 10^(-avgDigits)
}
```

I think if max(breaks) < max_x, then max_x must have more than `avgDigits` decimals. If it had fewer or the same, then the first rounding would have rounded the value to itself and `max(breaks) < max_x` would be false. 
`max_x` having more than `avgDigits` digits means that `max_x - max(breaks) < 10^(-avgDigits).

Reply:
i'm going to grab a snack and see if the above still makes sense after eating...

Reply:
so the math checks out, but i've since learned that even though we specify the digits in the arguments to `cut`, it's just using it as a guideline and will still do its own rounding. 
A proper fix of rewriting `cut` would take some time, so not something we can do for b58. I'll make a new issue for it so that we don't lose it. Bin labels and how they relate to the precise binWidth is a good data viz topic.

I'd vote to merge what we already have, we'll fix `cut` for b59, and also move to `numBins` for the map for b59.

Reply:
looks good!

I'd have done it with `while` rather than work out if the maths is correct, but it's dangerous!

```
while (max(breaks) < max_x) {
  breaks[length(breaks)] <- breaks[length(breaks)]  + 10^(-avgDigits)
}
```

Reply:
I know that's a bit sloppy.
Did you consider

```
if (max(breaks) < max_x) {
  breaks[length(breaks)] <- max_x + 10^(-avgDigits)
}
```
?

Reply:
ahh yes i did consider `max_x + 10^(-avgDigits)`, but then we would have the last bin not rounded like all the others

Reply:
Ah, yes, them's the breaks.